### PR TITLE
Power tooltip

### PR DESCRIPTION
**Pull request by DnAp**
I added information about using power in tooltip.

I don't know it is good idea or not.
It is my first experience in OpenRa.

### CODE REVIEW

**Comment by MustaphaTR** [comment]
In fact this is something already wanted as you can se from issue #6011. Not tested your code tho. A screenshot would be good.

File: OpenRA.Mods.Common/Traits/PowerTooltip.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	[Desc("Shown power info on the build palette widget.")]
+	public class PowerTooltipInfo : UpgradableTraitInfo
+	{
+		public override object Create(ActorInitializer init) { return new PowerTooltip(init.Self, this); }
+	}
+
+	public class PowerTooltip : IProvideTooltipInfo
+	{
+		readonly Actor self;
+		readonly PowerTooltipInfo info;
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/PowerTooltip.cs]: This is never used, and that makes the CIS fail the build.

File: OpenRA.Mods.Common/Traits/PowerTooltip.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		readonly PowerManager powerManager;
+
+		public PowerTooltip(Actor self, PowerTooltipInfo info)
+		{
+			this.self = self;
+			this.info = info;
+			powerManager = self.Owner.PlayerActor.Trait<PowerManager>();
+		}
+
+
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/PowerTooltip.cs]: Only one empty line, please.

File: OpenRA.Mods.Common/Traits/PowerTooltip.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+		public bool IsTooltipVisible(Player forPlayer)
+		{
+			return forPlayer == self.Owner;
+		}
+
+		public string TooltipText
+		{
+			get { return "Power Usage: " + powerManager.PowerDrained.ToString() + (powerManager.PowerProvided != 1000000 ? "/" + powerManager.PowerProvided.ToString() : ""); }
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/PowerTooltip.cs]: You should use "Power usage: {0}{1}...".F(...)

**Comment by teees** [comment]
Looks like this:
http://gfycat.com/CoarseThornyIbis

**Comment by Phrohdoh** [comment]
I like it, but in RA2 (as reference) each power provider displayed the amount of power they added to the overall power amount.

Currently this is just duplicate data of the already-available overall power usage.

**Comment by MustaphaTR** [comment]
> but in RA2 (as reference) each power provider displayed the amount of power they added to the overall power amount.

No, they doesn't they shows Power/Drain only too.

**Comment by atlimit8** [comment]
If this extra information is attached to power sources, then drain makes more sense than usage since usage implies that it is using rather than producing power which is obviously incorrect.
"Power Drain: 0/400"

This could be useful for mods without power information in the panel. However, I would prefer a "Power: x" where x is the current power output calculated by enabled `Power` traits and `ScalePowerWithHealth`; Although, if one is familiar with the rules, they could guess power output by upgrade overlays and the health bar.

Perhaps, a trait property such as `Total: true` or `TotalDrain: true` could be used for mods without power information in the panel, and the default could be used to tell players a structure's current power output.

**Comment by obrakmann** [comment]
Does this need to be a new trait? I think it would work just as well if you just made `Power` implement `IProvideTooltipInfo`. Modders would then control it using an enum in `PowerInfo` like this:

```yaml
Building:
    Power:
        ShowTooltipInfo: Always, Never, ConsumerOnly or ProducerOnly
```
Alternatively, just a boolean would probably be enough.

What do others think?

**Comment by atlimit8** [comment]
@obrakmann, actually, one actor can have multiple `Power` traits ("turbines"); if `Power` implemented `IProvideTooltipInfo`, then that info could be added multiple times.

**Comment by obrakmann** [comment]
True, hadn't thought of that. Carry on, then.

File: OpenRA.Mods.Common/Traits/PowerTooltip.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+		}
+
+		public bool IsTooltipVisible(Player forPlayer)
+		{
+			return forPlayer == self.Owner;
+		}
+
+		public string TooltipText
+		{
+			get {
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/PowerTooltip.cs]: Please put the brace on a new line:
```
get
{
	return ...;
}

**Comment by abcdefg30** [comment]
Looks good to me. :+1: / :white_check_mark: 

Could you squash the commits together, please?

**Comment by DnAp** [comment]
Done

File: OpenRA.Mods.Common/Traits/PowerTooltip.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+		public bool IsTooltipVisible(Player forPlayer)
+		{
+			return forPlayer == self.Owner;
+		}
+
+		public string TooltipText
+		{
+			get
+			{
+				return "Power Usage: {0}{1}".F(powerManager.PowerDrained.ToString(), powerManager.PowerProvided != 1000000 ? "/" + powerManager.PowerProvided.ToString() : "");
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/PowerTooltip.cs]: These `.ToString()`s shouldn't be needed?

File: OpenRA.Mods.Common/Traits/PowerTooltip.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ * see COPYING.
+ */
+#endregion
+
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Shown power info on the build palette widget.")]
+	public class PowerTooltipInfo : UpgradableTraitInfo
```
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/PowerTooltip.cs]: This inherits from `UpgradableTraitInfo`, but `PowerTooltip` does not inherit from `UpgradableTrait`. The simple solution is to replace `UpgradableTraitInfo` with `ITraitInfo` and remove `override` from `Create`.

File: OpenRA.Mods.Common/Traits/PowerTooltip.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+		public bool IsTooltipVisible(Player forPlayer)
+		{
+			return forPlayer == self.Owner;
+		}
+
+		public string TooltipText
+		{
+			get
+			{
+				return "Power Usage: {0}{1}".F(powerManager.PowerDrained, powerManager.PowerProvided != 1000000 ? "/" + powerManager.PowerProvided : "");
```
Comment (ScottNZ) [inline OpenRA.Mods.Common/Traits/PowerTooltip.cs]: I don't understand this `!= 1000000` bit. What are you trying to do?

Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/PowerTooltip.cs]: It is/was a magic value for the unlimited power cheat. It should be replaced with something like `self.Owner.PlayerActor.Trait<DeveloperMode>().UnlimitedPower`. Actually, `self.Owner.PlayerActor.Trait<DeveloperMode>()` should be cached. See OpenRA.Mods.Common/Traits/Power/Player/PowerManager.cs

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark:

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c82346ddedc96113571f4954df3c7c7ac2255f17...cb83922a1c24f1a5c68f2cef30bac91f136af3e4)

**Pull request merged**